src/mds/CInode.cc

```diff
@@ -5176,6 +5182,111 @@ void CInode::maybe_export_pin(bool update)
   }
 }
 
+void CInode::maybe_export_ephemeral_pin(bool update)
+{
+  //If both the configs aren't set then return
+  if (!mdcache->get_export_ephemeral_distributed_config()
+      && !mdcache->get_export_ephemeral_random_config())
+    return;
+
+  double export_ephemeral_random_pin = get_export_ephemeral_random_pin(false);
+  //bool export_ephemeral_distributed_pin = get_export_ephemeral_distributed_pin();
+
+  if (update)
+    dout(10) << "I'm here for the update" << dendl;
+
+  //Check if it's already ephemerally pinned
+  if ((is_export_ephemeral_random_pinned || is_export_ephemeral_distributed_pinned) && !update)
+      return;
+
+  if (mdcache->get_export_ephemeral_random_config()) {
+    if ((update || export_ephemeral_random_pin >=
+	  ceph::util::generate_random_number(0.0, 1.0))
+	&& is_export_ephemeral_distributed_pinned == false) {
```

Comment:
If we set random pin on a directory, all descendant directories under the directory will  be randomly distributed to all mds. This may create lots of subtrees.  lot of subtrees add lots of overhead on mds journal. I don't know what 's the use case of ephemeral_random_pin

Reply:
The whole subtree below it is supposed to be migrated to a random rank. It is hierarchical like export_pin. So the implementation is almost the same.

Reply:
export_ephemeral_random_pin is from  get_export_ephemeral_random_pin(), which return true if any ancestor dir has export_ephemeral_random_pin enabled .  This means inodes under an export_ephemeral_random_pin enabled directory will be randomly distributed

Reply:
In practice, `export_ephemeral_random_pin` will be something like 1% or maybe significantly less. It would be rare for a child directory to be ephemerlaly pinned.

Reply:
Iwhat benefits we can get from export_ephemeral_random_pin?

Reply:
Because the pinning is hierarchical, it's somewhat likely that a randomly ephemerally pinned cold directory will still be a parent of a hot directory.

What real gains we see from this is still unknown. I think it warrants research and experimentation after it's done.

Reply:
unless a hot directory is in very deep path, the hot directory is still unlikely to be distributed. I still don't think export_ephemeral_random_pin is useful.

Reply:
That may be, we'll see!